adapters/http-vertx-base/src/test/java/org/eclipse/hono/adapter/http/AbstractVertxBasedHttpProtocolAdapterTest.java

```diff
@@ -787,6 +792,114 @@ public void testUploadTelemetryWithTtdClosesCommandConsumerIfSenderCreationFaile
+    @Test
+    public void testUploadTelemetryWithDisabledGatewayMapping() {
+
+        // FIRST verify command consumer creation with a tenant without special configuration:
+        // GIVEN an adapter with a downstream telemetry consumer attached
+        final HttpServer server = getHttpServer(false);
+        final AbstractVertxBasedHttpProtocolAdapter<HttpProtocolAdapterProperties> adapter = getAdapter(server, null);
+        givenATelemetrySenderForOutcome(Future.succeededFuture());
+
+        final DeviceConnectionClient devConClient = mock(DeviceConnectionClient.class);
+        when(devConClient.setLastKnownGatewayForDevice(anyString(), any(), any())).thenReturn(Future.succeededFuture());
+        when(deviceConnectionClientFactory.getOrCreateDeviceConnectionClient(anyString())).thenReturn(Future.succeededFuture(devConClient));
+
+        final String tenantId = "tenant";
+        final String deviceId = "device";
+        final String gatewayId = "gw-1";
+        final DeviceUser deviceUser = new DeviceUser(tenantId, gatewayId);
+
+        when(commandConsumerFactory.createCommandConsumer(anyString(), anyString(), eq(gatewayId), any(Handler.class), any(Handler.class)))
+                .thenReturn(Future.succeededFuture(commandConsumer));
+
+        // and a device with a 'via' property
+        final JsonObject assertRegistrationResult = new JsonObject();
+        assertRegistrationResult.put(RegistrationConstants.FIELD_VIA, new JsonArray().add(gatewayId));
+        when(regClient.assertRegistration(anyString(), any(), any())).thenReturn(Future.succeededFuture(assertRegistrationResult));
+
+        // WHEN a device publishes a telemetry message with a TTD
+        final Buffer payload1 = Buffer.buffer("some payload");
+        final HttpServerResponse response1 = mock(HttpServerResponse.class);
+        final HttpServerRequest request1 = mock(HttpServerRequest.class);
+        when(request1.getHeader(eq(Constants.HEADER_TIME_TILL_DISCONNECT))).thenReturn("10");
+        final RoutingContext ctx1 = newRoutingContext(payload1, "text/plain", request1, response1);
+        when(ctx1.user()).thenReturn(deviceUser);
+        when(ctx1.addBodyEndHandler(any(Handler.class))).thenAnswer(invocation -> {
+            final Handler<Void> handler = invocation.getArgument(0);
+            handler.handle(null);
+            return 0;
+        });
+
+        adapter.uploadTelemetryMessage(ctx1, tenantId, deviceId);
+
+        // THEN the device receives a 202 response immediately
+        verify(response1).setStatusCode(202);
+        verify(response1).end();
+        // and the last known gateway has been set to the gateway
+        verify(devConClient).setLastKnownGatewayForDevice(eq(deviceId), eq(gatewayId), any());
+
+        // and the command consumer has been created with the gateway id
+        verify(commandConsumerFactory).createCommandConsumer(eq(tenantId), eq(deviceId), eq(gatewayId), any(Handler.class), any(Handler.class));
```

Comment:
since this is all about making sure that one variant of the method is invoked and not the other, I would suggest to also check that the other variant has **not** been invoked:
```
verify(commandConsumerFactory, never()).createCommandConsumer(anyString(), anyString(), any(Handler.class), any(Handler.class));
```
and do the same below for the other variant.
WDYT?

Reply:
Yes, that makes it more explicit. I've changed the method accordingly.